### PR TITLE
Support stable diffusion with dynamic shapes in DeepSparse

### PR DESCRIPTION
**Pull request by mwitiderrick**
```python
from optimum.deepsparse import  DeepSparseStableDiffusionPipeline
model_id = "stabilityai/stable-diffusion-2-1"
pipeline = DeepSparseStableDiffusionPipeline.from_pretrained(model_id,export=True)
prompt = "Surrealist painting of a floating island with giant clock gears, populated with mythical creatures."
outputs = pipeline(prompt=prompt, num_inference_steps = 75, guidance_scale = 7.5,height=512, width=768).images
outputs[0]
```
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205173540568853

### CODE REVIEW

File: optimum/deepsparse/__init__.py [hunk 1]
```diff
@@ -31,6 +31,18 @@
         "DeepSparseModelForSemanticSegmentation",
         "DeepSparseModelForTokenClassification",
     ],
+    "modeling_diffusion": [
+        "DeepSparseStableDiffusionXLImg2ImgPipeline",
+        "DeepSparseStableDiffusionXLPipeline",
+        "DeepSparseStabletableDiffusionXLPipelineBase",
```
Comment (mgoin) [inline optimum/deepsparse/__init__.py]: we still want to remove the XL pipelines, right?

File: tests/deepsparse/test_stable_diffusion.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
+#  Copyright 2021 The HuggingFace Team. All rights reserved.
```
Comment (mgoin) [inline tests/deepsparse/test_stable_diffusion.py:1]: this works currently on a few systems locally so we can land

File: optimum/deepsparse/modeling_diffusion.py [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+    DIFFUSION_MODEL_UNET_SUBFOLDER,
+    DIFFUSION_MODEL_VAE_DECODER_SUBFOLDER,
+    DIFFUSION_MODEL_VAE_ENCODER_SUBFOLDER,
+)
+
+from .modeling_base import DeepSparseBaseModel
+
+
+ONNX_WEIGHTS_NAME = "model.onnx"
+os.environ["NM_DISABLE_BATCH_OVERRIDE"] = "1"
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py]: move this within the compile function - we don't want to affect other code executing that might import this

File: optimum/deepsparse/modeling_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+        Args:
+            vaem_decoder_session (`ort.InferenceSession`):
+                The ONNX Runtime inference session associated to the VAE decoder.
+            text_encoder_session (`ort.InferenceSession`):
+                The ONNX Runtime inference session associated to the text encoder.
+            unet_session (`ort.InferenceSession`):
+                The ONNX Runtime inference session associated to the U-NET.
+            config (`Dict[str, Any]`):
+                A config dictionary from which the model components will be instantiated. Make sure to only load
+                configuration files of compatible classes.
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py]: Update this documentation

File: optimum/deepsparse/modeling_diffusion.py [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+            tokenizer (`CLIPTokenizer`):
+                Tokenizer of class
+                [CLIPTokenizer](https://huggingface.co/docs/transformers/v4.21.0/en/model_doc/clip#transformers.CLIPTokenizer).
+            scheduler (`Union[DDIMScheduler, PNDMScheduler, LMSDiscreteScheduler]`):
+                A scheduler to be used in combination with the U-NET component to denoise the encoded image latents.
+            feature_extractor (`Optional[CLIPFeatureExtractor]`, defaults to `None`):
+                A model extracting features from generated images to be used as inputs for the `safety_checker`
+            vae_encoder_session (`Optional[ort.InferenceSession]`, defaults to `None`):
+                The ONNX Runtime inference session associated to the VAE encoder.
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py]: ditto

File: optimum/deepsparse/modeling_diffusion.py [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+        for name in sub_models.keys():
+            self._internal_dict[name] = (
+                ("diffusers", "OnnxRuntimeModel") if sub_models[name] is not None else (None, None)
+            )
+        self._internal_dict.pop("vae", None)
+
+    def compile(self, height, width, batch_size=1):
+        unet_batch_size = 2
+        # GET THIS FROM THE UNET CONFIG
+        # https://github.com/huggingface/optimum/blob/e6f83c550316a0859bc7183d8bdca6a85fffede8/optimum/pipelines/diffusers/pipeline_stable_diffusion.py#L290
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py]: remove this debug comment

File: optimum/deepsparse/modeling_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+
+        # Modify config to keep the resulting model compatible with diffusers pipelines
+        for name in sub_models.keys():
+            self._internal_dict[name] = (
+                ("diffusers", "OnnxRuntimeModel") if sub_models[name] is not None else (None, None)
+            )
+        self._internal_dict.pop("vae", None)
+
+    def compile(self, height, width, batch_size=1):
+        unet_batch_size = 2
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py:148]: is there a batch size to set for the whole pipeline? how does this interact with the internal batch size - just a simple multiplication?

File: optimum/deepsparse/modeling_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+        # GET THIS FROM THE UNET CONFIG
+        # https://github.com/huggingface/optimum/blob/e6f83c550316a0859bc7183d8bdca6a85fffede8/optimum/pipelines/diffusers/pipeline_stable_diffusion.py#L290
+
+        # Create a dummy vae decoder in order to load config information
+        self.temp_vae_decoder = DeepSparseModelVaeDecoder(None, self)
+        if "block_out_channels" in self.temp_vae_decoder.config:
+            self.vae_scale_factor = 2 ** (len(self.temp_vae_decoder.config["block_out_channels"]) - 1)
+        else:
+            self.vae_scale_factor = 8
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py]: can you also put this in the `is_static` section or is there other state to maintain?

File: optimum/deepsparse/modeling_diffusion.py [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+            self.unet_engine = deepsparse.Engine(
+                model=str(self.unet_path), batch_size=unet_batch_size, input_shapes=unet_input_shapes
+            )
+            if os.path.exists(self.text_encoder_2_path):
+                self.text_encoder_2_engine = deepsparse.Engine(
+                    model=str(self.text_encoder_2_path), batch_size=batch_size, input_shapes=text_encoder_input_shapes
+                )
+
+        else:
+            logger.info("Compiling model with dynamic shapes, not recommended..")
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py]:
```suggestion
            logger.info("Compiling model with dynamic shapes, not recommended for performance..")
```

File: optimum/deepsparse/modeling_diffusion.py [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+    def to(self, device: str):
+        self._device = device.upper()
+        return self
+
+    @property
+    def device(self) -> str:
+        return self._device.lower()
+
+
+# TODO : Use ORTModelPart once IOBinding support is added
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py]:
```suggestion
```

File: optimum/deepsparse/modeling_diffusion.py [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+
+    CONFIG_NAME = "config.json"
+
+    def __init__(self, engine: deepsparse.Engine, parent_model: DeepSparseBaseModel, model_name: str = "encoder"):
+        self.engine = engine
+        self.parent_model = parent_model
+        self._model_name = model_name
+        config_path = parent_model.model_save_dir / model_name / self.CONFIG_NAME
+        self.config = self.parent_model._dict_from_json_file(config_path) if config_path.is_file() else {}
+        self.input_dtype = {"latent_sample": np.float64}
```
Comment (mgoin) [inline optimum/deepsparse/modeling_diffusion.py:413]: does this need to be in the base class? i don't see it used

File: optimum/deepsparse/__init__.py (hunk 1 above)
Comment (mwitiderrick) [reply optimum/deepsparse/__init__.py]: Right

File: optimum/deepsparse/modeling_diffusion.py (hunk 11 above)
Comment (mwitiderrick) [reply optimum/deepsparse/modeling_diffusion.py:413]: Which line exactly are you referring to here? I am certain line 418 will throw an error if removed and line 417 is used to get the model configuration.

File: optimum/deepsparse/modeling_diffusion.py (hunk 7 above)
Comment (mwitiderrick) [reply optimum/deepsparse/modeling_diffusion.py:148]: The whole pipeline works with a default batch size of 1. What do you mean by `internal batch size` ?

**Review by mgoin — APPROVED** [review]
Woo!

**Pull request merged**